### PR TITLE
Step Selector: Fix fob position rendering bug and remove CardFobAdaptor

### PR DESCRIPTION
**Pull request by japie1235813**
* Motivation for features / changes
Current we define a `CardFobController` and implements the methods at CardFobController level (e.g. `ScalarCardFobController`) and pass "this" to child component. However,  Angular will not detech the data changes within `ScalarCardFobController` (in our case, minMax and axisSize) because as far as it concerns, "this" does not change.
This causes the bug when minMax and axisSize are updated (changing window size, expand card group..) it does not cause card_fob_component to be re-rendered and the fob position is off.

* Technical description of changes
To handle this, we need to modify the method `getAxisPositionFromStep` which transform step to position, to two separate methods, getAxisPositionFromStartStep and getAxisPositionFromEndStep, and execute these two methods on `ScalarCardFobController` level, and feed the result to CardFobController. This way we are sure the input data is changed and Angular will detect the changes then re-render the CardFobController.

For the rest of CardFobAdaptor methods, they are used to get step and involving mouse movements. They are either do not have the same re-render requirements or the changes will cause timeSelection changes, which for sure will trigger child component re-rendering.
Therefore, I preserve this four methods and rename it to CardFobGetStepHelper and pass it to child component.
This PR does not include test changes and is used for discussion.

Update: we only keep getStepHigherThanPosition and getStepLowerThanPosition in a helper object and pass it as functions. The other 4 methods are used as passing values.

### CODE REVIEW

File: tensorboard/webapp/widgets/card_fob/card_fob_types.ts [hunk 1]
```diff
@@ -29,35 +29,24 @@ export enum AxisDirection {
 }
 
 /**
- * This class is intended to be implemented by the card that has a
+ * These helper functions are intended to be used in the card that has a
```
Comment (bmd3k) [inline tensorboard/webapp/widgets/card_fob/card_fob_types.ts:32]: "These helper functions are intended to be defined by the card that is a parent of CardFobControllerComponent." ?

File: tensorboard/webapp/widgets/card_fob/card_fob_controller_component.ts [hunk 2]
```diff
@@ -41,9 +45,12 @@ export class CardFobControllerComponent {
   @ViewChild('endFobWrapper') readonly endFobWrapper!: ElementRef;
   @Input() axisDirection!: AxisDirection;
   @Input() timeSelection!: TimeSelection;
-  @Input() cardAdapter!: CardFobAdapter;
+  @Input() cardFobHelper!: CardFobGetStepFromPositionHelper;
+  @Input() axisPositionFromStartStep!: number;
```
Comment (bmd3k) [inline tensorboard/webapp/widgets/card_fob/card_fob_controller_component.ts:49]: Consider this wordy but more accurate name: timeSelectionStartStepAxisPosition.
Similarly: timeSelectionEndStepAxisPosition.

or combine them into a single object TimeSelectionAxisPositions: {
  start: number;
  end: number;
}

(I started wondering what the difference is supposed to be between lowestStep/startStep and highestStep/endStep, then realized startStep and endStep were specifically related to TimeSelection).

File: tensorboard/webapp/widgets/histogram/histogram_card_fob_controller.ts [hunk 3]
```diff
@@ -24,30 +30,50 @@ import {TemporalScale} from './histogram_component';
...
+    getStepHigherThanAxisPosition:
+      this.getStepHigherThanAxisPosition.bind(this),
+    getStepLowerThanAxisPosition: this.getStepLowerThanAxisPosition.bind(this),
+  };
+
+  getAxisPositionFromStartStep() {
+    return this.temporalScale(this.timeSelection.start.step);
+  }
+
+  getAxisPositionFromEndStep() {
```
Comment (bmd3k) [inline tensorboard/webapp/widgets/histogram/histogram_card_fob_controller.ts:62]: Note:

@Input() axisPositionFromEndStep!: number;

... does not include `null` in the type. Unfortunately our Angular compiler does not catch this in the OSS code base but we should consider updating the return type to better reflect the possibility.

File: tensorboard/webapp/widgets/card_fob/card_fob_controller_test.ts [hunk 4]
```diff
@@ -40,22 +48,27 @@ class TestableComponent {
 
   @Input() axisDirection!: AxisDirection;
   @Input() timeSelection!: TimeSelection;
-  @Input() cardFobAdapter!: CardFobAdapter;
+  @Input() cardFobHelper!: CardFobGetStepFromPositionHelper;
   @Input() showExtendedLine?: Boolean;
 
   @Input() onTimeSelectionChanged!: (newTimeSelection: TimeSelection) => void;
   @Input() onTimeSelectionToggled!: () => void;
+  @Input() getAxisPositionFromStartStep!: () => number;
```
Comment (bmd3k) [inline tensorboard/webapp/widgets/card_fob/card_fob_controller_test.ts:56]: Can you change the types of these four new properties from function to plain-old number? And refactor the tests to reflect this?

File: tensorboard/webapp/widgets/card_fob/card_fob_controller_test.ts [hunk 5]
```diff
@@ -83,29 +96,37 @@ describe('card_fob_controller', () => {
...
     });
-    fixture.componentInstance.cardFobAdapter = cardFobAdapter;
+    fixture.componentInstance.cardFobHelper = cardFobHelper;
+    fixture.componentInstance.getAxisPositionFromStartStep =
+      getAxisPositionFromStartStepSpy;
+    fixture.componentInstance.getAxisPositionFromEndStep =
+      getAxisPositionFromEndStepSpy;
+    fixture.componentInstance.getHighestStep = getHighestStepSpy;
+    fixture.componentInstance.getLowestStep = getLowestStepSpy;
+    fixture.componentInstance.cardFobHelper = cardFobHelper;
```
Comment (bmd3k) [inline tensorboard/webapp/widgets/card_fob/card_fob_controller_test.ts:129]: Note: Duplicate of L122.

File: tensorboard/webapp/widgets/card_fob/card_fob_controller_test.ts (hunk 4 above)
Comment (bmd3k) [reply tensorboard/webapp/widgets/card_fob/card_fob_controller_test.ts:56]: I assume you could do the following on L123 to 128:
```

    fixture.componentInstance.axisPositionFromStartStep = input.timeSelection.start.step;
    fixture.componentInstance.axisPositionFromEndStep = input.timeSelection.end ? input.timeSelection.end.step : null;
    fixture.componentInstance.highestStep = 4;
    fixture.componentInstance.lowestStep = 0;
```

**Review by bmd3k — APPROVED** [review]

File: tensorboard/webapp/widgets/card_fob/card_fob_controller_test.ts (hunk 4 above)
Comment (japie1235813) [reply tensorboard/webapp/widgets/card_fob/card_fob_controller_test.ts:56]: With assigning to axisPositionFromStartStep directly, the tests are failing. (for example the first test below "vertical dragging" at line 166 and line 182 fails)
After debugging I found the reason is that although the initial value is taken from input.timeSelection, the timeSelection variable got updated from mousemove, but it never updates "`startStepAxisPosition`" and "`getCssTranslatePxForStartFob`" still returns the same initial number.

I think highestStep and lowestStep can be assigned from input but `axisPositionFromStartStep` and `axisPositionFromEndStep` should be kept as function and check the spy. 

Let me know if you have any other suggestions.

Comment (bmd3k) [reply tensorboard/webapp/widgets/card_fob/card_fob_controller_test.ts]: Oh I see. Yes, that makes sense! Thanks for explaining.

**Pull request merged**